mtags/src/main/scala-0.23/scala/scala/meta/internal/pc/CompilerInterfaces.scala

```diff
@@ -0,0 +1,94 @@
+package scala.meta.internal.pc
+
+import java.net.URI
+import java.nio.file.Paths
+import java.io.BufferedWriter
+import java.io.OutputStreamWriter
+import java.{util => ju}
+import scala.collection.JavaConverters._
+import scala.io.Codec
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.reporting.StoreReporter
+import dotty.tools.dotc.parsing.Parsers.Parser
+import dotty.tools.dotc.reporting.diagnostic.Message
+import dotty.tools.dotc.reporting.StoreReporter
+import dotty.tools.dotc.reporting.diagnostic.MessageContainer
+import dotty.tools.dotc.interfaces.Diagnostic
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.io.VirtualFile
+import org.eclipse.{lsp4j => l}
+
+// note(@tgodzik) the plan is to be able to move the methods here back to Dotty compiler
+// so that we can provide easier compatibility with multiple Scala 3 versions
+object CompilerInterfaces {
+
+  def parseErrors(
+      driver: InteractiveDriver,
+      uri: URI,
+      text: String
+  ): ju.List[l.Diagnostic] = {
+    val sourceFile = toSource(uri, text)
+    // we need to have a separate reporter otherwise errors are not cleared
+    val parsingReporter = new StoreReporter(null)
+    val freshContext = driver.currentCtx.fresh.setReporter(parsingReporter)
+    val parser = new Parser(sourceFile)(freshContext)
+    parser.parse()
+    val diags = parsingReporter.allErrors
+    diags.flatMap(diagnostic).asJava
+  }
+
+  private def diagnostic(mc: MessageContainer): Option[l.Diagnostic] =
+    if (!mc.pos.exists)
+      None // diagnostics without positions are not supported: https://github.com/Microsoft/language-server-protocol/issues/249
```

Comment:
Probably not ideal, but it _could_ depend on what the diagnostic is, but if it is relevant to an entire file for example, we could just detect it, give it a position at the beginning of the file and display it.

Reply:
Good idea, this message comes from Dotty though. Might have copied it with the code :O